File: gitgossipbackend/src/models/userModel.js
import mongoose from "mongoose";

/**
 * User Schema
 * -----------
 * This schema represents an application user.
 * Authentication is handled by Firebase,
 * and MongoDB is used to store additional user data.
 */
const userSchema = new mongoose.Schema(
    {
        /**
         * Firebase UID
         * - Comes from Firebase Authentication
         * - Acts as the primary identifier for a user
         */
        firebaseUid: {
            type: String,
            required: true,
            unique: true,
            index: true,
            immutable: true,
        },

        /**
         * User's display name
         */
        fullName: {
            type: String,
            trim: true,
        },

        /**
         * Unique username chosen by the user
         */

        username: {
            type: String,
            unique: true,
            sparse: true,
            lowercase: true,
            trim: true,
            minlength: 3,
            maxlength: 20,
            match: /^[a-z0-9_]+$/,


        },

        /**
         * User email address
         */
        email: {
            type: String,
            required: true,
            unique: true,
            lowercase: true,
            trim: true,
        },

        /**
         * User phone number
         */
        userPhoneNumber: {
            type: String,
            unique: true,
            sparse: true,
            match: /^\+[1-9]\d{7,14}$/,
        },


        /**
         * Short user bio
         */
        userBio: {
            type: String,
            sparse: true,
            trim: true,
        },

        /**
         * URL of user's profile picture
         */
        profilePic: {
            type: String,
        },

        /** 🔥 NEW: Skills */
        techStack: {
            type: [String],
            default: [],
            lowercase: true,
            index: true,
        },
        /**
         * Social media & portfolio links
         */
        socialLinks: {
            instagram: { type: String, trim: true },
            github: { type: String, trim: true, match: /^https?:\/\// },
            linkedin: { type: String, trim: true, match: /^https?:\/\// },
            portfolio: { type: String, trim: true, match: /^https?:\/\// },

        },
    },
    {
        /**
         * Automatically adds:
         * - createdAt
         * - updatedAt
         */
        timestamps: true,
    }
);




const User = mongoose.model("User", userSchema);
export default User;
/*
mongoose.model() = creates a link between schema and MongoDB collection.
export default = allows you to import and use it anywhere else in your app.
*/

File: gitgossipfrontend/lib/features/userProfile/models/user_model.dart
import 'package:gitgossip/features/userProfile/models/social_links_model.dart';

class UserModel {
  /// MongoDB document ID (_id)
  final String id;

  /// Firebase Authentication UID
  final String firebaseUid;

  final String? fullName;
  final String? username;
  final String email;

  /// Stored as String to support country codes (+91, etc.)
  final String? userPhoneNumber;
  final String? userBio;

  final String? profilePic;
  final List<String>? techStack;

  final SocialLinks? socialLinks;

  final DateTime? createdAt;
  final DateTime? updatedAt;

  UserModel({
    required this.id,
    required this.firebaseUid,
    this.fullName,
    this.username,
    required this.email,
    this.userPhoneNumber,
    this.userBio,
    this.profilePic,
    this.techStack,
    this.socialLinks,

    this.createdAt,
    this.updatedAt,
  });

  // ---------- FROM JSON ----------
  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'],
      firebaseUid: json['firebaseUid'],
      fullName: json['fullName'],
      username: json['username'],
      email: json['email'],
      userPhoneNumber: json['userPhoneNumber'],
      userBio: json["userBio"],
      profilePic: json['profilePic'],
      techStack: json['techStack'] != null
          ? List<String>.from(json['techStack'])
          : [],
      socialLinks: json['socialLinks'] != null
          ? SocialLinks.fromJson(json['socialLinks'])
          : null,

      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : null,
    );
  }

  // ---------- TO JSON ----------
  Map<String, dynamic> toJson() {
    return {
      "_id": id,
      "firebaseUid": firebaseUid,
      "fullName": fullName,
      "username": username,
      "email": email,
      "userPhoneNumber": userPhoneNumber,
      "userBio": userBio,
      "profilePic": profilePic,
      'techStack': techStack,
      "socialLinks": socialLinks?.toJson(),
    };
  }

  // ---------- COPY WITH ----------
  UserModel copyWith({
    String? id,
    String? firebaseUid,
    String? fullName,
    String? username,
    String? email,
    String? userPhoneNumber,
    String? userBio,
    String? profilePic,
    List<String>? techStack, // ✅ Corrected
    SocialLinks? socialLinks,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return UserModel(
      id: id ?? this.id,
      firebaseUid: firebaseUid ?? this.firebaseUid,
      fullName: fullName ?? this.fullName,
      username: username ?? this.username,
      email: email ?? this.email,
      userPhoneNumber: userPhoneNumber ?? this.userPhoneNumber,
      userBio: userBio ?? this.userBio,
      profilePic: profilePic ?? this.profilePic,
      techStack: techStack ?? this.techStack,
      socialLinks: socialLinks ?? this.socialLinks,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
